### PR TITLE
YarpPlugin: Automatic .ini files generation

### PR DESCRIPTION
**Pull request by drdanz**
This PR adds automatic `.ini` generation and installation to `yarp_prepare_plugin` and `yarp_install`.

This should fix a lot of issues caused by the duplicate information (already known by CMake).

While, for now, having several plugins in the same `.ini` file is still allowed, this patch enforces to have only one plugin per `.ini` file. It also enforces to have the same `name` and `part`, and therefore the `part` entries are not added to the file.

### CODE REVIEW

**Comment by traversaro** [comment]
That is great! 

A common failure mode for YARP devices is when the plugin name and the package name passed to `yarp_configure_external_installation` or `yarp_configure_plugins_installation` are equal, that results in `<something>.ini` being generate both for the package and the plugin, and one shadows/cancels the other in the installation. This is even more probably if no `<something>.ini` file is manually created by the user. Do you think it could make sense to add a check related to this in this PR, or it could make sense to have it in a separate PR?

**Comment by Nicogene** [comment]
If I understood correctly, the ini is generated only in `install` phase, in this way `make install` became mandatory to use the yarp plugins(devices, carriers).
I personally doesn't install, I use the `YARP_DATA_DIRS` var that points to the `build/share/yarp` folder.
Is there a possibility to generate it during the compile step?

**Comment by PeterBowman** [comment]
> If I understood correctly, the ini is generated only in `install` phase

`yarp_install` used to generate a copy of to-be-installed components in `build/share/` on CMake configure step, so I believe that should be fine.

**Comment by drdanz** [comment]
> If I understood correctly, the ini is generated only in install phase, in this way make install became mandatory to use the yarp plugins(devices, carriers).

The `.ini` file is actually generated by `yarp_add_plugin` and the magic is that a target property is set on the target. The property is then read by `yarp_install` and the file is both installed and copied in the build directory, in the expected position.

TL;DR This doesn't change anything for the user, both the build and install tree will have the file where it used to be before.

**Comment by drdanz** [comment]
> Do you think it could make sense to add a check related to this in this PR, or it could make sense to have it in a separate PR?

I'd go for a separate PR.
To be honest, I believe that the `.ini` file generated by `yarp_configure_plugins_installation` is completely useless, and I'd like to get rid of them. For example, this is the `yarp.ini` file.

```
###### This file is automatically generated by CMake.
[search yarp]
path "/home/ddomenichelli/Workspace/iit/build/yarp/lib/yarp"
extension ".so"
type "shared"
```

The only useful information here is `path`. The `.so` extension is platform dependent, and the `shared` type is obvious (the .ini files are not used for static plugins).

I'd like to change the `.ini` files generated by this patch to include the whole path, instead of just the library name, and get rid of these files (an alternative would be to include this information in the `path.d`)

This requires some changes in YARP though, it's not just a matter of CMake.

Also, there are a few extra things to consider:
 1. We will need to have different .ini files for debug and release for MSVC and Xcode, therefore `YARP_PREFIX_PATH` should probably change. On one side this will simplify the plugin search logic, but on the other side I consider it a "breaking change", since users will have to change the environment variables.
 1. These files will not be "portable" (AFAIK, the `yarp.ini` file is at the moment the only file that contains an absolute path). I'd like to find a way to make it portable.


Also, since with this patch these `.ini` files are finally completely hidden to the user, it's easier to change their names, therefore I'd like to replace the "plugin name" (e.g. `mjpeg`) with the "full plugin name" (`yarpcar_mjpeg.ini`). This should reduce the number of conflicts.

**Comment by traversaro** [comment]
> I'd like to change the `.ini` files generated by this patch to include the whole path, instead of just the library name, and get rid of these files (an alternative would be to include this information in the `path.d`)

The problem is probably already there for the existing `.ini` in  `yarp_configure_plugins_installation`, but this seems to be non-relocatable? How are we handling this in YARP/iCub Windows binaries? Plugins are broken in that case?

**Comment by traversaro** [comment]
> Also, since with this patch these `.ini` files are finally completely hidden to the user, it's easier to change their names, therefore I'd like to replace the "plugin name" (e.g. `mjpeg`) with the "full plugin name" (`yarpcar_mjpeg.ini`). This should reduce the number of conflicts. 

:+1:

**Comment by claudiofantacci** [comment]
If we plan to go for several PR for improving/updating plugin usage in YARP, it would also be the chance to refresh its [documentation](http://yarp.it/add_a_plugin.html) that, my personal two cents, is quite cryptic 😄

**Comment by drdanz** [comment]
> If we plan to go for several PR for improving/updating plugin usage in YARP, it would also be the chance to refresh its documentation that, my personal two cents, is quite cryptic smile

@claudiofantacci Cool, thanks for volunteering to do that! :rofl: :rofl: :rofl:

**Comment by claudiofantacci** [comment]
> @claudiofantacci Cool, thanks for volunteering to do that! 🤣 🤣 🤣

I'm totally fine about updating the documentation, provided that I have some support from experts 😛

**Comment by drdanz** [comment]
> The problem is probably already there for the existing `.ini` in `yarp_configure_plugins_installation`, but this seems to be non-relocatable? How are we handling this in YARP/iCub Windows binaries? Plugins are broken in that case?

At the moment, if I remember correctly, in the YARP Windows installer, the yarp.ini file is magically re-generated by NSIS during the installation. I don't remember about iCub, but there are several other relocation issues there, I think this is also handled by NSIS or other scripts.

**Comment by traversaro** [comment]
> At the moment, if I remember correctly, in the YARP Windows installer, the yarp.ini file is magically re-generated by NSIS during the installation.

Wow, I  was not aware of that.

**Comment by drdanz** [comment]
Just for reference, this patch introduced a few bugs, if you pulled after this, please update to the latest `devel`
